faunadb-android/src/test/java/com/faunadb/client/test/FaunaDBTest.java

```diff
@@ -6,37 +6,54 @@
 import com.faunadb.client.types.Value;
 import com.faunadb.client.types.Value.RefV;
 import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
 import com.google.common.util.concurrent.AsyncFunction;
 import com.google.common.util.concurrent.ListenableFuture;
 import org.junit.AfterClass;
 import org.junit.BeforeClass;
 
+import java.util.List;
+
 import static com.faunadb.client.query.Language.*;
 import static com.faunadb.client.types.Codec.REF;
 import static com.faunadb.client.types.Codec.STRING;
 import static com.faunadb.client.types.Value.NullV.NULL;
 import static com.google.common.base.Functions.constant;
 import static com.google.common.util.concurrent.Futures.*;
-import static java.lang.String.format;
 
-public class FaunaDBTest {
+public abstract class FaunaDBTest extends FaunaDBTestCommon {
```

Comment:
I liked the idea of passing functions so we don't bind ourselves to a specific interface. 

I think that each client specific test should be the one instantiating its client. Perhaps, it would be easier to understand if we design them as:
- `DslSpec`: abstract class that has protected and abstract `query` methods so we can keep schema definition and generic classes in it. I don't care if we don't use static setup/teardown. We could do a workaround with a boolean flag to only initialise the schema once per test execution. I know it's not ideal but, those tests are suppose to run only once per driver and I think we can reduce the overall complexity of our test setup if we do so. I don't foresee any race condition that could cause them to be flaky at the moment.
- `JavaDriverSpec`: extends `DslSpec`, instantiate java driver, implement its query methods, specific driver tests, and destroy its test database the tests run.
- `AndroidDriverSpec`: extends `DslSpec`, instantiate android driver, implement its query methods, specific driver tests, and destroy its test database the tests run.

Essentially, we could reduce the overall complexity if we get around the static setup/teardown by pushing them down on the class hierarchy and using a static flag when that is not possible.